File: katas/typescript/src/Patterns/builder/builder-good.ts
// Implementación del Patrón Builder: Construcción paso a paso con interfaz fluida
// ✅ Solución: Builder permite creación de objetos legible y flexible

// Clase Pizza con muchas propiedades opcionales
class Pizza {
  public size: string;
  public crust: string;
  public sauce: string;
  public cheese: string;
  public toppings: string[];
  public extraCheese: boolean;
  public spicyLevel: number;

  constructor(size: string, crust: string, sauce: string, cheese: string, toppings: string[], extraCheese: boolean, spicyLevel: number) {
    this.size = size;
    this.crust = crust;
    this.sauce = sauce;
    this.cheese = cheese;
    this.toppings = toppings;
    this.extraCheese = extraCheese;
    this.spicyLevel = spicyLevel;
  }

  public getDescription(): string {
    return `Pizza ${this.size} con masa ${this.crust}, salsa ${this.sauce}, queso ${this.cheese}, ingredientes: ${this.toppings.join(", ")}${this.extraCheese ? ", queso extra" : ""}, nivel picante: ${this.spicyLevel}/5`;
  }
}

// ✅ Builder con interfaz fluida
class PizzaBuilder {
  private size: string = "mediana";
  private crust: string = "regular";
  private sauce: string = "tomate";
  private cheese: string = "mozzarella";
  private toppings: string[] = [];
  private extraCheese: boolean = false;
  private spicyLevel: number = 0;

  public setSize(size: string): PizzaBuilder {
    this.size = size;
    return this; // ✅ Retorna 'this' para encadenamiento de métodos
  }

  public setCrust(crust: string): PizzaBuilder {
    this.crust = crust;
    return this;
  }

  public setSauce(sauce: string): PizzaBuilder {
    this.sauce = sauce;
    return this;
  }

  public setCheese(cheese: string): PizzaBuilder {
    this.cheese = cheese;
    return this;
  }

  public addTopping(topping: string): PizzaBuilder {
    this.toppings.push(topping);
    return this;
  }

  public withExtraCheese(): PizzaBuilder {
    this.extraCheese = true;
    return this;
  }

  public setSpicyLevel(level: number): PizzaBuilder {
    this.spicyLevel = level;
    return this;
  }

  public build(): Pizza {
    return new Pizza(
      this.size,
      this.crust,
      this.sauce,
      this.cheese,
      [...this.toppings], // Copia el array
      this.extraCheese,
      this.spicyLevel
    );
  }

  public reset(): PizzaBuilder {
    this.size = "mediana";
    this.crust = "regular";
    this.sauce = "tomate";
    this.cheese = "mozzarella";
    this.toppings = [];
    this.extraCheese = false;
    this.spicyLevel = 0;
    return this;
  }
}

// ✅ Beneficios del patrón builder:
console.log("=== Solución con Patrón Builder ===");

// Código legible y auto-documentado ✅
const margherita = new PizzaBuilder()
  .setSize("grande") // ✅ Claro qué es cada parámetro
  .setCrust("delgada")
  .setSauce("tomate")
  .setCheese("mozzarella")
  .addTopping("albahaca") // ✅ Fácil agregar ingredientes uno por uno
  .addTopping("tomates")
  .build();

// Fácil crear variaciones complejas ✅
const carnivora = new PizzaBuilder()
  .setSize("extra grande")
  .setCrust("gruesa")
  .setSauce("BBQ")
  .addTopping("pepperoni")
  .addTopping("salchicha")
  .addTopping("tocino")
  .addTopping("jamón")
  .withExtraCheese() // ✅ Características opcionales cuando se necesitan
  .setSpicyLevel(3)
  .build();

// Configuración mínima con valores por defecto ✅
const pizzaSimple = new PizzaBuilder().setSize("pequeña").addTopping("champiñones").build(); // ✅ Usa masa, salsa y queso por defecto

console.log("Margherita:", margherita.getDescription());
console.log("Carnívora:", carnivora.getDescription());
console.log("Pizza Simple:", pizzaSimple.getDescription());

export { Pizza, PizzaBuilder };

File: katas/python/Patterns/builder/builder_good.py
# Implementación del Patrón Builder: Construcción paso a paso con interfaz fluida
# ✅ Solución: Builder permite creación de objetos legible y flexible

from typing import List


# Clase Pizza con muchas propiedades opcionales
class Pizza:
    def __init__(
        self,
        size: str,
        crust: str,
        sauce: str,
        cheese: str,
        toppings: List[str],
        extra_cheese: bool,
        spicy_level: int,
    ):
        self.size = size
        self.crust = crust
        self.sauce = sauce
        self.cheese = cheese
        self.toppings = toppings
        self.extra_cheese = extra_cheese
        self.spicy_level = spicy_level

    def get_description(self) -> str:
        extra = ", queso extra" if self.extra_cheese else ""
        return (
            f"Pizza {self.size} con masa {self.crust}, salsa {self.sauce}, "
            f"queso {self.cheese}, ingredientes: {', '.join(self.toppings)}"
            f"{extra}, nivel picante: {self.spicy_level}/5"
        )


# ✅ Builder con interfaz fluida
class PizzaBuilder:
    def __init__(self):
        self.size = "mediana"
        self.crust = "regular"
        self.sauce = "tomate"
        self.cheese = "mozzarella"
        self.toppings: List[str] = []
        self.extra_cheese = False
        self.spicy_level = 0

    def set_size(self, size: str) -> "PizzaBuilder":
        self.size = size
        return self  # ✅ Retorna 'self' para encadenamiento de métodos

    def set_crust(self, crust: str) -> "PizzaBuilder":
        self.crust = crust
        return self

    def set_sauce(self, sauce: str) -> "PizzaBuilder":
        self.sauce = sauce
        return self

    def set_cheese(self, cheese: str) -> "PizzaBuilder":
        self.cheese = cheese
        return self

    def add_topping(self, topping: str) -> "PizzaBuilder":
        self.toppings.append(topping)
        return self

    def with_extra_cheese(self) -> "PizzaBuilder":
        self.extra_cheese = True
        return self

    def set_spicy_level(self, level: int) -> "PizzaBuilder":
        self.spicy_level = level
        return self

    def build(self) -> Pizza:
        return Pizza(
            self.size,
            self.crust,
            self.sauce,
            self.cheese,
            self.toppings.copy(),  # Copia la lista
            self.extra_cheese,
            self.spicy_level,
        )

    def reset(self) -> "PizzaBuilder":
        self.size = "mediana"
        self.crust = "regular"
        self.sauce = "tomate"
        self.cheese = "mozzarella"
        self.toppings = []
        self.extra_cheese = False
        self.spicy_level = 0
        return self


# ✅ Beneficios del patrón builder:
if __name__ == "__main__":
    print("=== Solución con Patrón Builder ===")

    # Código legible y auto-documentado ✅
    margherita = (
        PizzaBuilder()
        .set_size("grande")  # ✅ Claro qué es cada parámetro
        .set_crust("delgada")
        .set_sauce("tomate")
        .set_cheese("mozzarella")
        .add_topping("albahaca")  # ✅ Fácil agregar ingredientes uno por uno
        .add_topping("tomates")
        .build()
    )

    # Fácil crear variaciones complejas ✅
    carnivora = (
        PizzaBuilder()
        .set_size("extra grande")
        .set_crust("gruesa")
        .set_sauce("BBQ")
        .add_topping("pepperoni")
        .add_topping("salchicha")
        .add_topping("tocino")
        .add_topping("jamón")
        .with_extra_cheese()  # ✅ Características opcionales cuando se necesitan
        .set_spicy_level(3)
        .build()
    )

    # Configuración mínima con valores por defecto ✅
    pizza_simple = (
        PizzaBuilder().set_size("pequeña").add_topping("champiñones").build()
    )  # ✅ Usa masa, salsa y queso por defecto

    print("Margherita:", margherita.get_description())
    print("Carnívora:", carnivora.get_description())
    print("Pizza Simple:", pizza_simple.get_description())
